### ISSUE TITLE
[TF2.0] Gradients become None when using a SparseTensor kernel in a custom layer.

### ISSUE THREAD

**Issue by albertbuchard** [tensorflow/tensorflow#28966]
**System information**
- Have I written custom code: Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): MacOSX 10.14.5
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): v1.12.0-9492-g2c319fb415 2.0.0-alpha0
- Python version: 3.6
- CUDA/cuDNN version: running on mac/cpu
- GPU model and memory: running on mac/cpu

**Describe the current behavior**
In a DeepRL setting, using a DQN implementation in TF2. After a certain number of calls, the custom training loop with gradient tape is not able to compute gradient for the SparseMatrices used in the model (weights inside the sparse matrice), the gradient for biases is computed without any issue. 
The model can still be called, the weights are not zero. The gradients in _SparseTensorDenseMatMulGrad() are not None, and seem ok. That is, a_values_grad and b_grad are not None (also not None in _gradient_function)

`(None, a_values_grad, None, b_grad)`

But the output of `pywrap_tensorflow.TFE_Py_TapeGradient()` in `imperative_grad` has None for the weights in the SparseTensor. 

It is worth noting that if I do not call the model before the first training loop, the gradients are OK. Next loop they become None. 

**Describe the expected behavior**
Valid gradient for the weights in the SparseTensor whatever the number of calls done previously to the model.

**Code to reproduce the issue**
The model is quite complex. But the issue seem to come from one of the custom Layer which has a kernel defined like so:
```python 
kernel_weights = self.add_weight('kernel_weights',
                                         shape=[self.indices.shape[0]],
                                         initializer=self.kernel_initializer,
                                         regularizer=self.kernel_regularizer,
                                         constraint=self.kernel_constraint,
                                         dtype=self.dtype,
                                         trainable=self.trainable)

self.kernel = tf.SparseTensor(self.indices,
                                      kernel_weights,
                                      self.dense_shape)
```

And a call function defined as so:
```python
def call(self, x, **kwargs): 
        x = tf.transpose(tf.sparse.sparse_dense_matmul(self.kernel, x, adjoint_b=True))
        if self.use_bias:
            x += self.bias
        x = self.activation(x)
        return x
```

I tried to optimize the layer outside of the RL setting, with a dummy loss, and the issue remains. Here is the code for the custom loop:

```python
optimizer = Adam(learning_rate=1e-3)
x = tf.Variable(tf.ones((160, 41)))
data = Dataset.from_tensor_slices((x)).batch(20) 
for obs0 in data:
    with tf.GradientTape() as tape: 
        x = model(obs0) 
        
        loss_value = tf.reduce_mean(x, name='dummy_loss')
        
        grads = tape.gradient(loss_value, model.trainable_weights)
        # Here the gradient for the sparse kernel is None on the second loop
        #The gradient for the bias is fine
        
        optimizer.apply_gradients(zip(grads, model.trainable_weights))
```

**Other info / logs**
```
W0523 16:11:28.381706 4529878464 optimizer_v2.py:928] Gradients does not exist for variables ['custom_layer/kernel_weights:0'] when minimizing the loss.
```

**Comment by albertbuchard** [comment]
Ok so I solved the issue, at least the gradients are not None. I'm not sure if its a bug or not.

But if instead of storing the SparseTensor as a kernel i store the weights and build the sparse tensor at each call it work (albeit quite slow..):

build()
```python

        self.kernel_weights = self.add_weight('kernel_weights',
                                         shape=[self.indices.shape[0]],
                                         initializer=self.kernel_initializer,
                                         regularizer=self.kernel_regularizer,
                                         constraint=self.kernel_constraint,
                                         dtype=self.dtype,
                                         trainable=self.trainable)
```

And call()
```python
kernel = tf.SparseTensor(self.indices, self.kernel_weights, self.dense_shape)
x = tf.transpose(tf.sparse.sparse_dense_matmul(kernel, x, adjoint_b=True))        
``` 

Happy to close the issue, although im not sure if was the intended behaviour.

**Comment by ymodak** [comment]
Closing this issue since its resolved. Feel free to reopen if have any further questions. Thanks!

**Comment by kochbj** [comment]
Wanted to bump this up because I am running into the same issue. I have a couple of layers I have written with 
`tf.sparse.sparse_dense_matmul` that are not producing a gradient.
and also produce this warning:
`WARNING:tensorflow:Gradients do not exist for variables ['rgcn_autencoder/RGCN_Input0/graph_order_0_PcP_W_1:0', 'rgcn_autencoder/RGCN_Input0/graph_order_0_PbA_W_1:0',...when minimizing the loss.`
`
class RGCN_Input(Layer):
    def __init__(self,out_dim,graph_index,
                 relations= ['PcP','PbA','PiV','PuK','PbP','AwP','VpP','KpP','self'],
                 name="RGCN_Input",
                 activation=None,
                 use_bias=False,
                 kernel_initializer='glorot_uniform',
                 bias_initializer='zeros',
                 kernel_regularizer=None,
                 bias_regularizer=None,
                 **kwargs
                 ):
        super(RGCN_Input, self).__init__(name=name+str(graph_index),**kwargs)
        self.out_dim = out_dim
        self.relation = relations
        self.support = len(self.relation)
        self.graph_index = graph_index
        self.use_bias = use_bias
        self.activation = tf.keras.activations.get(activation)
        self.kernel_initializer = tf.keras.initializers.get(kernel_initializer)
        self.bias_initializer = tf.keras.initializers.get(bias_initializer)
        self.kernel_regularizer = tf.keras.regularizers.get(kernel_regularizer)
        self.bias_regularizer = tf.keras.regularizers.get(bias_regularizer)

    def build(self,input_shape):# define w & b
        self.kernel = tf.concat([
            self.add_weight(shape=(input_shape[1][1],self.out_dim),
                            name='graph_order_{}_{}_W_1'.format(self.graph_index,self.relation[i]),
                            trainable=True,
                            initializer=self.kernel_initializer,
                            regularizer=self.kernel_regularizer,
                           )
                            for i in range(self.support)]
            ,0)
        
        if self.use_bias:
            self.bias = self.add_weight(shape=(self.units,),
                                        initializer=self.bias_initializer,
                                        name='graph_order_{}_B_1'.format(self.graph_index),
                                        regularizer=self.bias_regularizer,
                                        )
        else: self.bias = None
        
        super(RGCN_Input, self).build(input_shape)

    def call(self,x): # define one r-gcn architecture
        assert isinstance(x, (list,tuple))
        adj_list,feature = x
        supports=[]
        if isinstance(adj_list[0], tf.SparseTensor):
            for i in range(self.support):
                supports.append(tf.sparse.sparse_dense_matmul(adj_list[i],feature))
        else:
            for i in range(self.support):
                supports.append(tf.matmul(adj_list[i],feature))
        support = tf.concat(supports,1) #squish them all together
        H = tf.matmul(support,self.kernel)

        if self.use_bias:
            H+=self.bias

        
        if self.activation is not None:
            H = self.activation(H)
        
        return H
    
    def compute_output_shape(self, input_shape):
        return (input_shape[1][0], self.output_dim)
`
